lib/bindings/python/tests/conftest.py

```diff
@@ -13,31 +13,347 @@
+                match = re.search(r":(\d+)$", address)
+                if match:
+                    etcd_client_port = int(match.group(1))
+                    print(f"ETCD bound to client port: {etcd_client_port}")
+                    break
+        except (json.JSONDecodeError, ValueError):
+            continue
+
+    if etcd_client_port is None:
+        etcd.terminate()
+        etcd.wait()
+        shutil.rmtree(nats_data_dir, ignore_errors=True)
+        shutil.rmtree(etcd_data_dir, ignore_errors=True)
+        raise RuntimeError("Failed to discover ETCD client port from logs")
+
+    # Now start NATS with retry logic (up to 5 attempts due to race condition)
+    max_nats_retries = 5
+    nats_server = None
+    nats_port = None
+    last_error = None
+
+    for attempt in range(max_nats_retries):
+        try:
+            nats_port = get_free_port()
+            print(
+                f"Attempt {attempt + 1}: Starting NATS on port {nats_port}, data dir: {nats_data_dir}"
+            )
+
+            nats_server = subprocess.Popen(
+                ["nats-server", "-js", "-p", str(nats_port), "-sd", str(nats_data_dir)],
+                stderr=subprocess.PIPE,
+            )
+
+            # Give NATS a moment to bind to the port
+            time.sleep(0.1)
+
+            # Check if NATS failed to start
+            if nats_server.poll() is not None:
+                stderr = (
+                    nats_server.stderr.read().decode() if nats_server.stderr else ""
+                )
+                if "address already in use" in stderr.lower():
+                    print(f"NATS port {nats_port} already in use, retrying...")
+                    time.sleep(0.1)
+                    continue
+                etcd.terminate()
+                etcd.wait()
+                shutil.rmtree(nats_data_dir, ignore_errors=True)
+                shutil.rmtree(etcd_data_dir, ignore_errors=True)
+                raise RuntimeError(f"NATS failed to start: {stderr}")
+
+            # Success - NATS started
+            break
+
+        except Exception as e:
+            last_error = e
+            print(f"Attempt {attempt + 1} failed: {e}")
+            if attempt < max_nats_retries - 1:
+                time.sleep(0.2)
+            else:
+                etcd.terminate()
+                etcd.wait()
+                shutil.rmtree(nats_data_dir, ignore_errors=True)
+                shutil.rmtree(etcd_data_dir, ignore_errors=True)
+                raise RuntimeError(
+                    f"Failed to start NATS after {max_nats_retries} attempts: {last_error}"
+                )
+
+    # Set environment variables for the runtime to use
+    os.environ["NATS_SERVER"] = f"nats://localhost:{nats_port}"
+    os.environ["ETCD_ENDPOINTS"] = f"http://localhost:{etcd_client_port}"
+
+    return nats_server, etcd, nats_port, etcd_client_port, nats_data_dir, etcd_data_dir
+
+
+@pytest.fixture(scope="module", autouse=True)
+def nats_and_etcd():
```

Comment:
https://github.com/ai-dynamo/dynamo/blob/d809906e62840e5a566c69bb364ececf56dba11b/tests/conftest.py#L218 

We also have this being used for other tests - can we combine?